### ISSUE TITLE
Externally control menu visibility

### ISSUE THREAD

**Issue by mbirnios** [ericgio/react-bootstrap-typeahead#325]
Hi there,

Would it be possible to consider adding a prop like "showMenuOnFocus" (default=true) that if set to false, would cause the menu NOT to dropdown when the input get focus? In other words, the menu appears when typing.

Our use case is that we have the typeahead above a grid. We are autocompleting usual search terms. When searching by clicking enter, the grid is loaded, but the typeahead still has the focus and the menu box is constantly hiding part of the grid.

![image](https://user-images.githubusercontent.com/30596501/37842436-dec30180-2ea0-11e8-87a1-cfcd6e7be639.png)

We tried a lot of workaround, with no success. We could move the focus to somewhere else, but then the user has to re-click on the typeahead to trigger new searches. We played a lot with OnChange and OnKeyDown, trying to force this behavior (using algo the blur() public method)... but it causes other side effects.

Thanks!

**Comment by ericgio** [comment]
Did you try the `minLength` prop? It specifies the number of characters a user needs to enter before displaying the dropdown.

You can try it out in [this example](http://ericgio.github.io/react-bootstrap-typeahead/#behaviors) (click the third checkbox).

**Comment by mbirnios** [comment]
Thank you! Yes, we have minLength = 2. But the user search typing 4 or 5 characters. Then our grid refresehs with the correct results... but the Typeahead remains open, obstructing the view of the grid.

We played with removing **{ showMenu: true }** form the **_handleFocus** function, and also, calling"**onHide()**" at onKeyDown handler that processes the RETURN key. That seemed to work like a charm!

Another thing we noticed is that even if "**submitFormOnEnter**" is set to {**True**}, if you select an option from the menu with they keyboard (by pressing enter), that doesn't submit the form right away. It just select the option and closes the menu, and you need to hit ENTER again. We applied a workaround for that too (so the user hits enter upon the menu option, and that submits the form right away).

If you see any value in these change, I can try to submit a PR for your review.

**Comment by ericgio** [comment]
I still don't understand the issue you're having and the desired behavior. Is there some example I can look at? The menu should close when a selection is made, so I'm not sure why it's staying open for you.

**Comment by mbirnios** [comment]
Thanks Eric! I'll try to come up with something. I agree that it's hard to explain... but let me try again. Using your "asynch" search from Github as an example.

1) You type, the menu appears, you select an option and hit ENTER.
2) The option gets selected and the menu is closed.
3) At that moment, we trigger a search to fill up a grid.
4) During the fetch, the screen gets blocked (we're using Block UI).
5) So the autocomplete loses focus.
6) Once the search finishes, the autocomplete gains focus again --> thus showing the menu again.

That's why I was willing to do two things:
1) Prevent the menu from opening by just focusing over.
2) Have "submitFormOnEnter" to also submit when you select a menu option (currently it just selects that option and closes the menu, and you need to hit ENTER again to actually do the submit).

I'll try to do some Fiddle or something.

**Comment by ericgio** [comment]
> Once the search finishes, the autocomplete gains focus again --> thus showing the menu again

Why does the typeahead need to re-gain focus?

**Comment by mbirnios** [comment]
All right, thanks @ericgio ! And great work by the way.

**Comment by xyogol** [comment]
Hello, the use case I am talking about is: on desktop version of the website we have a simple input field with Typeahead. And in tablet we would like to have the search input in the modal with all the menu items listed when modal is opened, so user could select something from the list without having the need to type or focus the input.

**Comment by dfagan2** [comment]
@ericgio - Similar desire as @xyogol from my implementation: a way to have the menu show at all times "underneath" the input. So if not an `alwaysOpen` prop of some sort to control the overlay, a way to designate a target container into which to render the menu (instead of an overlay)

**Comment by dfagan2** [comment]
@xyogol, @ericgio - I opened a pull request that helped me achieve what we've been mentioning: https://github.com/ericgio/react-bootstrap-typeahead/pull/406

**Comment by ericgio** [comment]
I think controlling menu visibility via props is reasonable. Here are some considerations I can think of, let me know if anyone has thoughts:

- There are various conditions internally where the menu is either shown or hidden. I think if the visibility is controlled by a prop (let's call it the `open` prop), then that prop should override all other conditions.
- Logically, I'd say the allowed values for the prop would be `undefined` (default), `true` (open), or `false` (closed).
- In addition to and controlled `open` prop, I think it would make sense to have an uncontrolled `defaultOpen` prop that renders the component with the menu open by default, but follows normal behaviors thereafter.

**Comment by ericgio** [comment]
`open` and `defaultOpen` props were added in v3.3.0
